plugins/image-source/obs-slideshow-mk2.c

```diff
@@ -0,0 +1,1192 @@
+#define T_TR_CUT                       T_TR_("Cut")
+#define T_TR_FADE                      T_TR_("Fade")
+#define T_TR_SWIPE                     T_TR_("Swipe")
+#define T_TR_SLIDE                     T_TR_("Slide")
+
+/* clang-format on */
+
+extern void image_source_preload_image(void *data);
+
+/* ------------------------------------------------------------------------- */
+
+struct image_file_data {
+	char *path;
+};
+
+struct source_data {
+	size_t slide_idx;
+	const char *path;
+	obs_source_t *source;
+};
+
+typedef DARRAY(struct image_file_data) image_file_array_t;
+
+enum behavior {
+	BEHAVIOR_STOP_RESTART,
+	BEHAVIOR_PAUSE_UNPAUSE,
+	BEHAVIOR_ALWAYS_PLAY,
+};
+
+#define SLIDE_BUFFER_COUNT 5
+
+struct active_slides {
+	struct circlebuf prev;
+	struct circlebuf next;
+	struct source_data cur;
+};
+
+struct slideshow_data {
+	struct active_slides slides;
+	image_file_array_t files;
+	float slide_time;
+	uint32_t tr_speed;
+	const char *tr_name;
+	bool manual;
+	bool randomize;
+	bool loop;
+	bool restart_on_activate;
+	bool pause_on_deactivate;
+	bool restart;
+	bool hide;
+	bool use_cut;
+	bool paused;
+	bool stop;
+	calldata_t cd;
+	float elapsed;
+	enum behavior behavior;
+
+	enum obs_media_state state;
+};
+
+struct slideshow {
+	obs_source_t *source;
+
+	struct slideshow_data data;
+	os_task_queue_t *queue;
+	obs_source_t *transition;
+	uint32_t cx;
+	uint32_t cy;
+
+	obs_hotkey_id play_pause_hotkey;
+	obs_hotkey_id restart_hotkey;
+	obs_hotkey_id stop_hotkey;
+	obs_hotkey_id next_hotkey;
+	obs_hotkey_id prev_hotkey;
+};
+
+static void set_media_state(void *data, enum obs_media_state state)
+{
+	struct slideshow *ss = data;
+	ss->data.state = state;
+}
+
+static enum obs_media_state ss_get_state(void *data)
+{
+	struct slideshow *ss = data;
+	return ss->data.state;
+}
+
+static inline obs_source_t *get_transition(struct slideshow *ss)
+{
+	return obs_source_get_ref(ss->transition);
+}
+
+static inline void free_source_data(struct source_data *sd)
+{
+	obs_source_release(sd->source);
+}
+
+static void free_active_slides(struct active_slides *slides)
+{
+	while (slides->prev.size) {
+		struct source_data file;
+		circlebuf_pop_front(&slides->prev, &file, sizeof(file));
+		free_source_data(&file);
+	}
+
+	while (slides->next.size) {
+		struct source_data file;
+		circlebuf_pop_front(&slides->next, &file, sizeof(file));
+		free_source_data(&file);
+	}
+
+	free_source_data(&slides->cur);
+
+	circlebuf_free(&slides->prev);
+	circlebuf_free(&slides->next);
+}
+
+static inline void free_slideshow_data(struct slideshow_data *ssd)
+{
+	free_active_slides(&ssd->slides);
+	for (size_t i = 0; i < ssd->files.num; i++)
+		bfree(ssd->files.array[i].path);
+	calldata_free(&ssd->cd);
+	da_free(ssd->files);
+}
+
+static size_t random_file(struct slideshow_data *ssd, size_t slide_idx)
+{
+	size_t next = slide_idx;
+
+	if (ssd->files.num > 1) {
+		while (next == slide_idx)
+			next = (size_t)rand() % ssd->files.num;
```

Comment:
Should be resolved now. Check fixup commit.

Reply:
I opted for:
```c
(~files + 1) % files
```
rather than using negate because negate causes a build error on windows (using negate on an unsigned value)